lib/seek/list_sorter.rb

```diff
@@ -38,6 +39,82 @@ class ListSorter
       created_at_desc: { title: 'Creation date (Descending)', order: 'created_at DESC' },
       position_asc: { title: 'Position (Ascending)', order: 'position' },
       position_desc: { title: 'Position (Descending)', order: 'position DESC' },
+      downloads_desc: { title: 'Downloads (Descending)', order: '--downloads_desc',
+                        relation_proc: -> (items) { # Sorts by number of downloads, descending
+                          #### Using Active Record
+                          # This section **modifies** "items" relation so that it includes a new column "downloads"
+                          alog=ActivityLog.all.where(action: 'download',activity_loggable_type: items.first.class.name)
+                          downloads=alog.select("activity_loggable_id AS #{items.table.name}_id, COUNT(activity_loggable_id) AS downloads").group(:activity_loggable_id)
+                          items._select!('*', 'd.downloads').joins!("LEFT OUTER JOIN (#{downloads.to_sql}) d ON #{items.table.name}.id = d.#{items.table.name}_id")
+
+                          #### Using Arel
+                          # This section builds the equivalent arel_table to provide the corresponding arel_field
+                          items_a=items.arel_table
+                          alog_a=ActivityLog.arel_table
+                          downloads=alog_a.project(alog_a[:activity_loggable_id].as("log_id"), alog_a[:activity_loggable_id].count.as("downloads"))
+                                        .where(alog_a[:action].eq('download').and(alog_a[:activity_loggable_type].eq(items.first.class.name)))
+                                        .group(:activity_loggable_id).as('downloads')
+                          joined=items_a.project(items_a[Arel.star], downloads[:downloads]).outer_join(downloads).on(items_a[:id].eq(downloads[:log_id])).as('d')
+
+                          case ActiveRecord::Base.connection.instance_values["config"][:adapter]
```

Comment:
This all seems very complex and long, and will be difficult to revisit and maintain in the future. Needs simplifying or breaking down. It could also be inconsistent with the `download_counts`, as it isn't include the `no_spider ` scope ( which needs improving, but will be more of a reason to include when it has been).

I wonder if there might be a way of changing the `count_actions ` queries in https://github.com/seek4science/seek/blob/master/lib/seek/stats/activity_counts.rb to a scope, and then chaining them together.

A possibility may be to split the query up rather than do one big join.

Another option may be to just sort programmatically, and then turn back into a relation (as is done with --relevence), doing this with all Datafiles on the FAIRDOMHub takes about 30 seconds but that could be an extreme case